utxo/src/cache.rs

```diff
@@ -13,202 +13,35 @@
-            UtxoSource::BlockChain(_) => false,
-            UtxoSource::MemPool => true,
-        }
-    }
-
-    fn blockchain_height(&self) -> Result<BlockHeight, Error> {
-        match self {
-            UtxoSource::BlockChain(h) => Ok(*h),
-            UtxoSource::MemPool => Err(crate::Error::NoBlockchainHeightFound),
-        }
-    }
-}
-
-/// The Unspent Transaction Output
-#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
-pub struct Utxo {
-    output: TxOutput,
-    is_block_reward: bool,
-    /// identifies whether the utxo is for the blockchain or for mempool.
-    source: UtxoSource,
-}
-
-impl Utxo {
-    pub fn new(output: TxOutput, is_block_reward: bool, height: BlockHeight) -> Self {
-        Self {
-            output,
-            is_block_reward,
-            source: UtxoSource::BlockChain(height),
-        }
-    }
-
-    /// a utxo for mempool, that does not need the block height.
-    pub fn new_for_mempool(output: TxOutput, is_block_reward: bool) -> Self {
-        Self {
-            output,
-            is_block_reward,
-            source: UtxoSource::MemPool,
-        }
-    }
-
-    pub fn is_block_reward(&self) -> bool {
-        self.is_block_reward
-    }
-
-    pub fn source_height(&self) -> &UtxoSource {
-        &self.source
-    }
-
-    pub fn output(&self) -> &TxOutput {
-        &self.output
-    }
-
-    pub fn set_height(&mut self, value: UtxoSource) {
-        self.source = value
-    }
-}
-
-pub trait UtxosView {
-    /// Retrieves utxo.
-    fn utxo(&self, outpoint: &OutPoint) -> Option<Utxo>;
-
-    /// Checks whether outpoint is unspent.
-    fn has_utxo(&self, outpoint: &OutPoint) -> bool;
-
-    /// Retrieves the block hash of the best block in this view
-    fn best_block_hash(&self) -> Id<GenBlock>;
-
-    /// Estimated size of the whole view (None if not implemented)
-    fn estimated_size(&self) -> Option<usize>;
-
-    fn derive_cache(&self) -> UtxosCache;
-}
+use std::{
+    collections::BTreeMap,
+    fmt::{Debug, Formatter},
+};
 
 #[derive(Clone)]
 pub struct ConsumedUtxoCache {
-    container: BTreeMap<OutPoint, UtxoEntry>,
-    best_block: Id<GenBlock>,
-}
-
-pub trait FlushableUtxoView {
-    /// Performs bulk modification
-    fn batch_write(&mut self, utxos: ConsumedUtxoCache) -> Result<(), Error>;
-}
-
-// flush the cache into the provided base. This will consume the cache and throw it away.
-// It uses the batch_write function since it's available in different kinds of views.
-pub fn flush_to_base<T: FlushableUtxoView>(cache: UtxosCache, base: &mut T) -> Result<(), Error> {
-    base.batch_write(cache.consume())
+    pub(crate) container: BTreeMap<OutPoint, UtxoEntry>,
+    pub(crate) best_block: Id<GenBlock>,
 }
 
 #[derive(Clone)]
 pub struct UtxosCache<'a> {
     parent: Option<&'a dyn UtxosView>,
     current_block_hash: Id<GenBlock>,
-    utxos: BTreeMap<OutPoint, UtxoEntry>,
-    //TODO: do we need this?
+    pub(crate) utxos: BTreeMap<OutPoint, UtxoEntry>,
```

Comment:
I think that it is worth putting a comment (maybe even todo?..) here that this `pub(crate)` is only needed for tests.